### ISSUE TITLE
Strong Monad with single parameterized remote monad

### ISSUE THREAD

**Issue by jtdawso** [ku-fpg/remote-monad#16]
To make the remote monad be able to do Haxl things we made the remote monad take a single parameter instead of the Command and Procedure. But since we need that split for the strong monad we had to introduce the `KnownResult` class (src/Control/Remote/Util.hs).  If we call `knownResult` we will get a `Just` with the static result or a `Nothing` letting us know that we have to go remote to get the result. 

The problem that we have is at the following lines of code
https://github.com/ku-fpg/remote-monad/blob/master/src/Control/Remote/Monad.hs#L140-L143

If we encounter something that has a static result (a `Command`) then we want to add the command to the queue and return the static result so we can proceed. But unfortunately, it seems that the type checker is getting upset that it can't tell that the `a` from the `knownResult` call is the same `a` that is in the type of the User datatype.

Any thoughts to get this working @andygill @RyanGlScott @roboguy13?

**Comment by RyanGlScott** [comment]
I'm not at all familiar with the finer details of this library, but I have battled typecheckers before, so perhaps I can be of assistance there. However, I'm not yet able to reproduce the problem you're experiencing. I tried uncommenting the `runStrongMonad` function but was greeted with this error:

```
src/Control/Remote/Monad.hs:141:64: error:
    Not in scope: data constructor ‘Strong.Primitive’
    Module ‘Control.Remote.Packet.Strong’ does not export ‘Primitive’.
    |
141 |           modify $ (\ (HStrongPacket cs) -> HStrongPacket (cs (Strong.Primitive p) ))
    |                                                                ^^^^^^^^^^^^^^^^
```

Do you know how to get past this hurdle?

**Comment by jtdawso** [comment]
I'm thinking our natural transformation stuff can be of assistance here, if we hide the a in the class we might be able to wrap and unwrap the natural transformation to make the type checker happy. I'm trying that now. If you would like to work in parallel, feel free to go for it!
For your issue use Strong.Command instead of Primitive

**Comment by RyanGlScott** [comment]
Hm. I still feel like I'm not quite at the point where you are yet. After changin `Strong.Primitive` to `Strong.Command`, I get this:

```
src/Control/Remote/Monad.hs:141:64: error:
    • Couldn't match expected type ‘StrongPacket prim x0’
                  with actual type ‘StrongPacket prim () -> StrongPacket prim ()’
    • Probable cause: ‘Command’ is applied to too few arguments
      In the first argument of ‘cs’, namely ‘(Command p)’
      In the first argument of ‘HStrongPacket’, namely ‘(cs (Command p))’
      In the expression: HStrongPacket (cs (Command p))
    • Relevant bindings include
        cs :: StrongPacket prim ~> StrongPacket prim
          (bound at src/Control/Remote/Monad.hs:141:38)
        p :: prim a (bound at src/Control/Remote/Monad.hs:138:22)
        go :: RemoteApplicative prim a
              -> MaybeT (StateT (HStrongPacket prim) m) a
          (bound at src/Control/Remote/Monad.hs:137:5)
        rf :: StrongPacket prim ~> m
          (bound at src/Control/Remote/Monad.hs:117:20)
        runStrongMonad :: (StrongPacket prim :~> m)
                          -> RemoteMonad prim :~> m
          (bound at src/Control/Remote/Monad.hs:117:1)
    |
141 |           modify $ (\ (HStrongPacket cs) -> HStrongPacket (cs (Strong.Command p) ))
    |                                                                ^^^^^^^^^^^^^^^^
```

This seems to suggest that `Strong.Command` isn't being given enough arguments. Indeed, its type suggests that it takes two:

https://github.com/ku-fpg/remote-monad/blob/260b951463daf29328595acce772f80a8d28a427/src/Control/Remote/Packet/Strong.hs#L23

But we're only giving it one, right?

Maybe it would help it you could prepare a branch that demonstrates the type error you're battling.

**Comment by jtdawso** [comment]
Sorry,
The line should be:
          modify $ (\ (HStrongPacket cs) -> HStrongPacket (cs . Strong.Command p))

and I've been toying around with passing `a` or `()` for the Command packet definition so that should be the only other change you need

**Comment by jtdawso** [comment]
There is a line that is missing as well for go2

    go2 (Appl app)   = go app

**Comment by RyanGlScott** [comment]
Thanks @jtdawso, that brings me up to speed.

Forgive me for asking what might be a silly question, but why are we passing `p` to `Command` in the first place? After all, we can't know statically that `knownResult` is going to return something of type `()`, so perhaps `Command` is the wrong choice here?

Or maybe I'm misunderstanding what `knownResult` is doing. _Is_ it supposed to be the case that `knownResult` should return something of type `()` here? If so, there are ways we could convince the typechecker of this fact. (But I want to make sure that this is in fact your intention.)

**Comment by jtdawso** [comment]
@RyanGlScott `knownResult` is the way that we can distinguish `Commands` from `Procedures` when using a single type to parameterize `RemoteMonad` .   Does that answer your question?

**Comment by jtdawso** [comment]
No luck on my side of things of pushing the following through: 
 `class KnownResult f where
  knownResult :: f :~> Maybe
  knownResult = wrapNT (\_ -> Nothing)`

**Comment by RyanGlScott** [comment]
Hm. I thought I had something, but ultimately, it seems that you'll get blocked by the fact that your state is an `HStrongPacket prim`—i.e., `forall x. StrongPacket prim x -> StrongPacket prim x`. There's simply no way to force `x` to be `()` with this overly general type, due to the way rank-_n_ types work. You could perhaps make this work by changing `HStrongPacket`, but I don't feel qualified to make such a change.

I'm afraid that's about all of the input I can offer on the matter—investigating deeper would require more intricate knowledge of `remote-monad`'s internals than I possess.
